### PR TITLE
Converting HTML Readme to Markdown

### PR DESCRIPTION
**Pull request by KunalGehlot**
This is to make the project more accessible to Github users.

Issue: #10

### CODE REVIEW

**Comment by KunalGehlot** [comment]
Please note that the contents of the README have **not** been altered except for a few broken internal links while converting to markdown.

File: README.md [hunk 1]
```diff
@@ -0,0 +1,438 @@
+# PoDoFo Readme
+
+## PoDoFo
+
+- [PoDoFo Readme](#podofo-readme)
+  - [PoDoFo](#podofo)
```
Comment (ccreutzi) [inline README.md:6]: These two lines seem unnecessary, can we just remove them (and ideally move the other lines to the left)?

File: README.md [hunk 2]
```diff
@@ -0,0 +1,438 @@
...
+- **podofotxtextract** - A tool that extracts all text from a PDF file. Works only for simple PDFs at the moment.
+- **podofouncompress** - Removes all compression filters from a PDF file. This is useful for debugging existing PDF files. You might also find PoDoFoBrowser, distributed separately, to be useful if you're using this tool a lot.
+
+Additionally there is the external tool PoDoFoBrowser which is not included in this package, but can be downloaded from the PoDoFo webpage. PoDoFoBrowser is a Qt application for browsing and modifying the objects in a PDF file, examining the structure of PDF files, and manipulating PDF streams. It is very useful if you want to look on the internal structure of PDF files.
+
+As of version 0.7 PoDoFo is available for Unix, Mac OS X and Windows platforms.
+
+## Where can I can get it?
+
+PoDoFo is available on the internet: [podofo.sf.net](http://podofo.sf.net)
```
Comment (ccreutzi) [inline README.md:71]: I appreciate not mixing the format transition with content changes, but we probably want to point at https://github.com/podofo/podofo/ now.

File: README.md [hunk 3]
```diff
@@ -0,0 +1,438 @@
...
+
+If your application or library uses Visual Studio, CMake can be used to build a Visual Studio project you can use. Alternately, if you are using MinGW for your existing code, CMake can make you a MinGW makefile to build a mingw compatible version of PoDoFo. You must build PoDoFo with the same compiler as your application. See [CMake builds on Windows](#cmake-builds-on-windows)
+
+Mac OS X users will need to use CMake to generate a conventional UNIX build with Makefiles, or may generate an XCode project with CMake. See the [Mac OS X section](#cmake-builds-on-mac-os-x) below, and the [generic instructions for cmake builds](#installation-with-cmake)
+
+### STLPort support
+
+Don't know what STLPort is? Ignore it and skip this section. Packagers should not enable STLPort.
+
+If your project uses STLPort rather than your platform/toolset's native STL, you must configure PoDoFo to use STLPort as well. Use the -DUSE\_STLPORT=1 argument to configure. Do not enable STLPort for podofo unless you use it in your project; you will get an incompatible C++ STL leading to link errors, runtime crashes, or other problems.
```
Comment (ccreutzi) [inline README.md:107]: Should -DUSE… be in backticks?

File: README.md [hunk 4]
```diff
@@ -0,0 +1,438 @@
...
+
+It is unnecessary and unwise to build PoDoFo against STLPort on any current major platform unless you have a specific reason to do so.
+
+### Boost support
+
+PoDoFo can make limited use of the Boost libraries if they are available. There is no need to use them, but some things may work more smoothly, and more use is expected in the future. If PoDoFo is built to use Boost, your application must also have Boost's headers (boost graph libraries) on its search path. Set -DWANT\_BOOST:BOOL=1 to enable Boost support.
```
Comment (ccreutzi) [inline README.md:113]: Should -DWANT… be in backticks?

File: README.md [hunk 5]
```diff
@@ -0,0 +1,438 @@
...
+
+You can use the CMake variable `CMAKE_BUILD_TYPE` to control the type of build. The main values supported are `DEBUG` and `RELEASE`. The default is DEBUG. Set the build type with the CMake argument:  
+  `-DCMAKE_BUILD_TYPE=DEBUG`  
+   or  
+  `-DCMAKE_BUILD_TYPE=RELEASE`  
+as appropriate.
+
+You can control where the files are installed with \`make install' with `-DCMAKE_INSTALL_PREFIX=/path/to/install/dir`
+
+All instructons below use out-of-tree builds (recommended). To clean up an out-of-tree build, just delete the build directory, as no files are touched within the source directory.
```
Comment (ccreutzi) [inline README.md:139]: Typo: instructions

File: README.md [hunk 6]
```diff
@@ -0,0 +1,438 @@
...
+Linux and UNIX users should be able to build PoDoFo by cd'ing into the PoDoFo checkout or unpacked tarball directory (here assumed to be named "podofo-src") then running the build commands shown below. The CMake command below will install into $HOME/podofo to avoid needing root priveleges for installation, but you can change the destination to wherever you want or remove the install prefix setting to use the default.
+
+To build and install:
+
+mkdir ../podofo-build
+cd ../podofo-build
+cmake -G "Unix Makefiles" -DCMAKE\_INSTALL\_PREFIX="$HOME/podofo" ../podofo-src
+make
+make install
+
```
Comment (ccreutzi) [inline README.md:154]: code block, should have ```

File: README.md [hunk 7]
```diff
@@ -0,0 +1,438 @@
...
+To see detailed compiler output, use:
+
+make VERBOSE=1
```
Comment (ccreutzi) [inline README.md:157]: code block, should have ```

File: README.md [hunk 8]
```diff
@@ -0,0 +1,438 @@
...
+
+If you're working on a 64 bit linux that has a /usr/lib64 directory, add `-DWANT_LIB64:BOOL=TRUE` to the CMake command line.
+
+If you need to specify additional paths to search for libraries, set the `CMAKE_INCLUDE_PATH` and `CMAKE_LIBRARY_PATH` environment variables or set them on the command line:
+
+cmake -G "Unix Makefiles" ../podofo-src \\
+ -DCMAKE\_INCLUDE\_PATH=/usr/sfw/include \\
+ -DCMAKE\_LIBRARY\_PATH=/usr/sfw/lib
```
Comment (ccreutzi) [inline README.md:165]: code block, should have ```

File: README.md [hunk 9]
```diff
@@ -0,0 +1,438 @@
...
+
+If you wish to generate only a static library or only a shared library, set the `PODOFO_BUILD_SHARED` and/or `PODOFO_BUILD_STATIC` flags:
+
+cmake -G "Unix Makefiles" ../podofo-src \\
+ -DCMAKE\_INCLUDE\_PATH=/usr/sfw/include \\
+ -DCMAKE\_LIBRARY\_PATH=/usr/sfw/lib \\
+ -DPODOFO\_BUILD\_SHARED:BOOL=TRUE \\
+ -DPODOFO\_BUILD\_STATIC:BOOL=FALSE
```
Comment (ccreutzi) [inline README.md:173]: code block, should have ```

File: README.md [hunk 10]
```diff
@@ -0,0 +1,438 @@
...
+
+You will need freetype and fontconfig. It's possible to omit the use of libjpeg (see the platform-independent documentation) but by default libjpeg will also be required. If you have these libraries already (say, from fink or DarwinPorts) you can skip the following section and update the `CMAKE_INCLUDE_PATH` and `CMAKE_LIBRARY_PATH` arguments appropriately.
+
+##### Dependencies
+
+I'll assume you're installing podofo and any required libraries into $HOME/libs . Adjust to taste, but keep it consistent.
+
+The easiest way to get any required libraries is to use [MacPorts](http://macports.org/) to install the libjpeg, fontconfig, and freetype libraries. Once you've installed MacPorts, just run:
+
+/opt/local/bin/port install fontconfig freetype jpeg tiff lua
```
Comment (ccreutzi) [inline README.md:193]: code block, should have ```

File: README.md [hunk 11]
```diff
@@ -0,0 +1,438 @@
...
+ cmake -G "Unix Makefiles" \\
+  -DWANT\_FONTCONFIG:BOOL=TRUE \\
+  -DCMAKE\_INSTALL\_PREFIX=/opt/podofo \\
+  -DCMAKE\_INCLUDE\_PATH=/opt/local/include \\
+  -DCMAKE\_LIBRARY\_PATH=/opt/local/lib \\
+  ../podofo
+ make
+ sudo mkdir /opt/podofo
+ sudo chown $USER /opt/podofo
+ make install
```
Comment (ccreutzi) [inline README.md:212]: code block, should have ```

File: README.md [hunk 12]
```diff
@@ -0,0 +1,438 @@
...
+
+For Homebrew (package manager for Mac OS X) it is quite simple to install podofo. Freetype2 and zlib should be installed on Mac OS X by default.
+
+brew install fontconfig
+brew install libjpeg
+brew install libtiff
+cd ((directory of podofo))
+cmake -G "Unix Makefiles"
+make
+make install
```
Comment (ccreutzi) [inline README.md:226]: code block, should have ```

File: README.md [hunk 13]
```diff
@@ -0,0 +1,438 @@
...
+
+###### libjpeg
+
+MinGW users can download the latest libjpeg installer from [GnuWin32](http://gnuwin32.sf.net)
+
+.
+
+For Visual Studio you will need to build libjpeg with your version of Visual Studio. Download the libjpeg sources from [http://www.ijg.org/](http://www.ijg.org/) and unpack them into a working directory. I'll assume you've used `C:\Developer\jpeg`. Once the sources are unpacked, start a visual studio command prompt and cd into `c:\developer\jpeg` then run `copy jconfig.vc jconfig.h` then `nmake /f makefile.vc /a`
+
+.
```
Comment (ccreutzi) [inline README.md:248]: Both of these look like the full stop/period should be part of the preceding paragraph.

File: README.md [hunk 14]
```diff
@@ -0,0 +1,438 @@
...
+
+If you use a DLL build of libjpeg remember that you must ship the DLL with your application. During testing the libjpeg dll must be in the same directory as the PoDoFo-using executable or the directory containing it must be on your system PATH. These requirements do not apply to the (recommended) static library libjpeg.
+
+##### Building PoDoFo on Windows
+
+Building PoDoFo on Windows can be done using MinGW (a minimalist gcc-based compliler environment for Windows) or Visual Studio. Other methods may work but have not been tested.
+
+CMake 2.6.x is required for Windows. You can download it from [cmake.org](http://www.cmake.org/HTML/Download.html)
+
+.
```
Comment (ccreutzi) [inline README.md:258]: Looks like the full stop/period should be part of the preceding paragraph.

File: README.md [hunk 15]
```diff
@@ -0,0 +1,438 @@
...
+
+Because the C++ ABIs of most of the win32 compilers are incompatible, you must build PoDoFo with the same compiler and version that you will use to build the programs linked to PoDoFo. Failure to follow this restriction will result in link errors at best, and bizarre runtime failures at worst.
+
+On Windows, if you are linking against a shared (DLL) build of PoDoFo you MUST define the preprocessor macro `USING_SHARED_PODOFO` when including any podofo headers. Failure to do so will result in link time and/or runtime errors. Similarly, defining it when linking with a static PoDoFo can cause problems.
+
+On Windows, PoDoFo may be built as either a shared or static library. Building both is not supported. By default only the shared library will be built. If you want a static library, just disable generation of the shared library with the extra argument to cmake:
+
+ -DPODOFO\_BUILD\_SHARED=FALSE
```
Comment (ccreutzi) [inline README.md:266]: code block, should have ```

File: README.md [hunk 16]
```diff
@@ -0,0 +1,438 @@
...
+For release you should generally build your own copy of these libraries unless there is a well known and high quality pre-built version, like there is for zlib. If you have built your own libjpeg you can improve PoDoFo's use of it a bit by passing `-DJPEG_COMPATIBLE_RUNTIME` to CMake to tell PoDoFo it's safe not to use its' workarounds for incompatible runtimes.
+
+##### Handling library naming on win32
+
+Especially on Windows it is also common for custom built libraries to have different names to those you might download as pre-built copies. CMake won't be able to find them if they're called something else unless you tell it. Use these variables to tell CMake what names to look for a library under:
+
+- FREETYPE\_LIBRARY\_NAMES\_DEBUG, FREETYPE\_LIBRARY\_NAMES\_RELEASE and FREETYPE\_LIBRARY\_NAMES
+- TIFF\_LIBRARY\_NAMES\_DEBUG, TIFF\_LIBRARY\_NAMES\_RELEASE and TIFF\_LIBRARY\_NAMES
+- LIBJPEG\_LIBRARY\_NAMES\_DEBUG, LIBJPEG\_LIBRARY\_NAMES\_RELEASE and LIBJPEG\_LIBRARY\_NAMES
+- ZLIB\_LIBRARY\_NAMES\_DEBUG, ZLIB\_LIBRARY\_NAMES\_RELEASE, ZLIB\_LIBRARY\_NAMES
```
Comment (ccreutzi) [inline README.md:279]: These should be in backticks, I believe.

File: README.md [hunk 17]
```diff
@@ -0,0 +1,438 @@
...
+To build PoDoFo with MinGW, you'll naturally need MinGW. The author recommends installing Qt 4 from Trolltech, which has a well packaged version of MinGW and is also useful for some PoDoFo tools like PoDoFoBrowser.
+
+Once MinGW is set up, make sure that the MinGW "bin" directory is on your PATH, and be sure to set `CMAKE_INCLUDE_PATH` and `CMAKE_LIBRARY_PATH` such that CMake can find the headers and .lib files for the libraries PoDoFo requires. The GnuWin32 library packages from [http://gnuwin32.sf.net/](http://gnuwin32.sf.net) are known to work with PoDoFo, so installing zlib, freetype, and libjpeg from there should do the trick.
+
+To configure and build PoDoFo with a default GnuWin32 install and with MinGW already on your PATH:
+
+md ..\\podofo-debug
+cd ..\\podofo-debug
+cmake -G "MinGW Makefiles" ..\\podofo-src -DCMAKE\_INCLUDE\_PATH=c:\\progra~1\\gnuwin32\\include -DCMAKE\_LIBRARY\_PATH=c:\\progra~1\\gnuwin32\\lib -DPODOFO\_BUILD\_SHARED:BOOL=FALSE
+mingw32-make
```
Comment (ccreutzi) [inline README.md:294]: code block, should have ```

File: README.md [hunk 18]
```diff
@@ -0,0 +1,438 @@
...
+If you built a debug freetype, you will also need to copy vc90.pdb (assuming you're using VC++ 9, otherwise the name will vary) from objs\\debug\_mt\\vc90.pdb to objs\\win32\\vc2008\\ so that VC++ can find Freetype's debug info.
+
+The visual studio build of PoDoFo has two stages - first, CMake is used to generate a Visual Studio solution, then the solution is built in VC++. I prefer to create a short `build.cmd` file for the CMake command since on Windows it tends to get rather long. Here's an example that assumes the `build.cmd` file is in an empty directory with `..\podofo-src` being the relative path to the PoDoFo sources.
+
+del cmakecache.txt
+set FTDIR=C:\\developer\\freetype-2.3.5
+set FTLIBDIR=C:\\developer\\freetype-2.3.5\\objs\\win32\\vc2008
+set JPEGDIR=C:\\Developer\\jpeg
+set ZLIBDIR=C:\\developer\\zlib123-dll
+cmake -G "Visual Studio 9 2008" ..\\podofo-src -DCMAKE\_INCLUDE\_PATH="%FTDIR%\\include;%JPEGDIR%\\include;%JPEGDIR%;%ZLIBDIR%\\include" -DCMAKE\_LIBRARY\_PATH="%FTLIBDIR%;%FTDIR%\\lib;%JPEGDIR%\\lib;%JPEGDIR%;%ZLIBDIR%\\lib" -DPODOFO\_BUILD\_SHARED:BOOL=FALSE -DFREETYPE\_LIBRARY\_NAMES\_DEBUG=freetype239MT\_D -DFREETYPE\_LIBRARY\_NAMES\_RELEASE=freetype239MT
```
Comment (ccreutzi) [inline README.md:315]: code block, should have ```

File: README.md [hunk 19]
```diff
@@ -0,0 +1,438 @@
...
+
+Note that CMake should automatically find your Visual Studio install, so you shouldn't need any special settings unless you have an unusual setup or more than copy of the same version of Visual Studio installed.
+
+You can download the free Visual C++ Express Edition 9 from [Microsoft](http://www.microsoft.com/express/).
+
+**PROBLEMS?**
+
+If you have problems, try deleting your build directory then re-running the commands above after running:
+
+set PATH=%SystemRoot%;%SystemRoot%\\SYSTEM32
```
Comment (ccreutzi) [inline README.md:327]: code block, should have ```

File: README.md [hunk 20]
```diff
@@ -0,0 +1,438 @@
...
+
+When PoDoFo reaches 1.x versions it is expected that API changes will be less frequent. At this point, though, they're an unavoidable part of improving and completing the library.
+
+If you are using CMake in your project and you choose to bundle a copy of the PoDoFo sources, with svn:externals or otherwise, you can quite easily build a private copy of PoDoFo as part of your app's normal build process. Have a look at FindLIBPODOFO.cmake from PoDoFoBrowser to see how to do that, or just copy that file.
+
+### Compiling and Linking Against PoDoFo
+
+PoDoFo's headers are designed to avoid exposing its library dependencies, so you do not need the headers for Freetype etc on the include path. However, you MUST link your application to podofo's required libraries on most platforms, eg on UNIX:
+
+    -lpodofo -lfreetype -lfontconfig -ljpeg -lz
```
Comment (ccreutzi) [inline README.md:357]: code block, should have ```

File: README.md [hunk 21]
```diff
@@ -0,0 +1,438 @@
...
+
+Several preprocessor defines are used inside PoDoFo. This section tries to document all of them. These defines are set automatically by the CMake build system, so usually you do not have to worry about them.
+
+**Define**
+
+**Meaning**
+
+`DEBUG`
+
+Indicates a debug build of PoDoFo.
```
Comment (ccreutzi) [inline README.md:377]: The formatting isn't great. I'd suggest we at least leave out the (pseudo-)headers, and ideally would use a [definition list](https://www.markdownguide.org/extended-syntax/#definition-lists).

**Review by ccreutzi — CHANGES_REQUESTED** [review]
This is a great improvement, thanks for the PR. If you're willing, I'd like to ask for a few more tweaks before merging. In particular, the code blocks are currently formatted as running text without their line breaks, I tried to mark all of them.

**Comment by ceztko** [comment]
I just merged the PR with Christopher suggestions, thank you. I would like to add that while this work is very welcome, I can't guarantee I will spend enough time on integrating the content from here when pushing [pdfmm](https://github.com/pdfmm/pdfmm) (which already had a Markdown readme) into PoDoFo. Also in a sense this README is too much verbose for today standards, so we'll see. Anyway it's a very useful work and it will be easier to source from here once PoDoFo-next is up.

**Pull request closed without merging**